File: Windows/OpenIoT.App/OpenIoT.Lib/Board/Models/BoardProperty.cs
using OpenIoT.Lib.Tools.Utils;

namespace OpenIoT.Lib.Board.Models
{
    public class BoardProperty
    {
        public int index;
        public object value;
        public int size;
        public int semantic;
        public int flags;
        public BoardPropertyType type;

        public static int Flag_Read = 1 << 0;
        public static int Flag_Write = 1 << 1;
        public static int Flag_Subscribed = 1 << 7;

        public static int Flag_Default = Flag_Read;

        public BoardProperty()
        {
        }

        public BoardProperty(int index, int value, int semantic, int flags)
        {
            this.Init(index, ByteUtils.FromInt32(value), 0, BoardPropertyType.Bool, semantic, flags);
        }

        public BoardProperty(int index, float value, int semantic, int flags)
        {
            this.Init(index, ByteUtils.FromFloat(value), 0, BoardPropertyType.Bool, semantic, flags);
        }

        public BoardProperty(int index, bool value, int semantic, int flags)
        {
            this.Init(index, new byte[] { ByteUtils.FromBool(value) }, 0, BoardPropertyType.Bool, semantic, flags);
        }

        public BoardProperty(int index, byte[] data, int dataOffset, BoardPropertyType dataType, int semantic, int flags)
        {
            this.Init(index, data, dataOffset, dataType, semantic, flags);
        }

        public BoardProperty(BoardProperty pOriginal, Object pValue)
        {
            this.index = pOriginal.index;
            this.semantic = pOriginal.semantic;
            this.flags = pOriginal.flags;
            this.type = pOriginal.type;
            this.size = pOriginal.size;

            if (this.type == BoardPropertyType.Data)
                this.SetValue((byte[])pValue, 0);
            else
                this.value = pValue;
        }

        private static readonly int[] sizes = { 0, 0, 4, 4, 1, 0 };

        public int Init(int index, byte[] data, int dataOffset, BoardPropertyType dataType, int semantic, int flags)
        {
            int offset = 0;

            this.index = index;
            this.semantic = semantic;
            this.flags = flags;
            this.type = dataType;


            switch (dataType)
            {
                case BoardPropertyType.Integer:
                    {
                        this.value = ByteUtils.ToInt32(data, dataOffset + offset);
                        this.size = sizes[(int)this.type];
                        offset += this.size;
                        break;
                    }
                case BoardPropertyType.Float:
                    {
                        this.value = ByteUtils.ToFloat(data, dataOffset + offset);
                        this.size = sizes[(int)this.type];
                        offset += this.size;
                        break;
                    }
                case BoardPropertyType.Bool:
                    {
                        this.value = ByteUtils.ToBool(data, dataOffset + offset);
                        this.size = sizes[(int)this.type];
                        offset += this.size;
                        break;
                    }
                case BoardPropertyType.Data:
                    {
                        this.size = data[dataOffset + offset];
                        offset++;
                        this.value = new byte[this.size];
                        ArrayUtils.Copy(data, dataOffset +  offset, (byte[])this.value, 0, this.size);
                        offset += this.size;
                        break;
                    }
            }

            return offset;
        }

        public bool IsWriteable()
        {
            return (this.flags & Flag_Write) != 0;
        }

        public bool IsReadable()
        {
            return (this.flags & Flag_Read) != 0;
        }

        public bool sSubscribed()
        {
            return (this.flags & Flag_Subscribed) != 0;
        }

        public int Size()
        {
            return this.size;
        }

        public void SetValue(object value)
        {
            switch (this.type)
            {
                case BoardPropertyType.Integer:
                    this.value = (int)value;
                    break;

                case BoardPropertyType.Float:
                    this.value = (float)value;
                    break;

                case BoardPropertyType.Bool:
                    this.value = (bool)value;
                    break;

                case BoardPropertyType.Data:
                    ArrayUtils.Copy((byte[])value, 0, (byte[])this.value, 0, this.size);
                    break;
            }
        }

        public int SetValue(byte[] data, int dataOffset)
        {
            int offset = 0;

            switch (this.type)
            {
                case BoardPropertyType.Integer:
                    {
                        this.value = ByteUtils.ToInt32(data, dataOffset);
                        offset += this.Size();
                        break;
                    }

                case BoardPropertyType.Float:
                    {
                        this.value = ByteUtils.ToFloat(data, dataOffset);
                        offset += this.Size();
                        break;
                    }

                case BoardPropertyType.Bool:
                    {
                        this.value = ByteUtils.ToBool(data, dataOffset);
                        offset += this.Size();
                        break;
                    }

                case BoardPropertyType.Data:
                    {
                        this.size = data[dataOffset];
                        offset++;
                        this.value = new byte[this.size];
                        ArrayUtils.Copy(data, dataOffset +  1, (byte[])this.value, 0, size);
                        offset += this.size;
                        break;
                    }
            }

            return offset;
        }

        public int getValue(byte[] data, int offset)
        {
            switch (this.type)
            {
                case BoardPropertyType.Integer:
                    {
                        return ByteUtils.FromInt32((int)this.value, data, offset);
                    }

                case BoardPropertyType.Float:
                    {
                        return ByteUtils.FromFloat((float)this.value, data, offset);
                    }

                case BoardPropertyType.Bool:
                    {
                        return ByteUtils.FromBool((bool)this.value, data, offset);
                    }

                case BoardPropertyType.Data:
                    {
                        int size = this.Size();
                        data[offset] = (byte)size;
                        ArrayUtils.Copy(data, offset, (byte[])this.value, offset + 1, size);
                        return 1 + size;
                    }
            }

            return 0;
        }

        public int GetInt()
        {
            return (int)this.value;
        }
        public float GetFloat()
        {
            return (float)this.value;
        }
        public bool GetBool()
        {
            return (bool)this.value;
        }

        public string GetString()
        {
            return StringUtils.BytesToHex((byte[])this.value);
        }
    }

}

File: Andriod/Constructor/construction/src/main/java/com/palitri/openiot/construction/framework/board/models/BoardProperty.java
package com.palitri.openiot.construction.framework.board.models;

import com.palitri.openiot.construction.framework.tools.utils.ArrayUtils;
import com.palitri.openiot.construction.framework.tools.utils.ByteUtils;
import com.palitri.openiot.construction.framework.tools.utils.StringUtils;

public class BoardProperty {
    public int index;
    public Object value;
    public int size;
    public int semantic;
    public int flags;
    public BoardPropertyType type;

    public static final int Flag_Read = 1 << 0;
    public static final int Flag_Write = 1 << 1;
    public static final int Flag_Subscribed = 1 << 7;

    public BoardProperty()
    {
    }

    public BoardProperty(int index, int value, int semantic, int flags)
    {
        this.Init(index, ByteUtils.fromInt32(value), 0, BoardPropertyType.Bool, semantic, flags);
    }

    public BoardProperty(int index, float value, int semantic, int flags)
    {
        this.Init(index, ByteUtils.fromFloat(value), 0, BoardPropertyType.Bool, semantic, flags);
    }

    public BoardProperty(int index, boolean value, int semantic, int flags)
    {
        this.Init(index, new byte[] { ByteUtils.fromBool(value) }, 0, BoardPropertyType.Bool, semantic, flags);
    }

    public BoardProperty(int index, byte[] data, int dataOffset, BoardPropertyType dataType, int semantic, int flags)
    {
        this.Init(index, data, dataOffset, dataType, semantic, flags);
    }

    public BoardProperty(BoardProperty pOriginal, Object pValue)
    {
        this.index = pOriginal.index;
        this.semantic = pOriginal.semantic;
        this.flags = pOriginal.flags;
        this.type = pOriginal.type;
        this.size = pOriginal.size;

        if (this.type == BoardPropertyType.Data)
            this.setValue((byte[])pValue, 0);
        else
            this.value = pValue;
    }

    public int Init(int index, byte[] data, int dataOffset, BoardPropertyType dataType, int semantic, int flags)
    {
        int offset = 0;

        this.index = index;
        this.semantic = semantic;
        this.flags = flags;
        this.type = dataType;

        switch (dataType) {
            case Integer: {
                this.value = new Integer(ByteUtils.toInt32(data, dataOffset + offset));
                this.size = this.type.GetStaticSize();
                offset += this.size;
                break;
            }
            case Float: {
                this.value = new Float(ByteUtils.toFloat(data, dataOffset + offset));
                this.size = this.type.GetStaticSize();
                offset += this.size;
                break;
            }
            case Bool: {
                this.value = new Boolean(ByteUtils.toBool(data, dataOffset + offset));
                this.size = this.type.GetStaticSize();
                offset += this.size;
                break;
            }
            case Data: {
                this.size = data[dataOffset + offset];
                offset++;
                this.value = new byte[this.size];
                ArrayUtils.Copy(data, dataOffset +  offset, this.value, 0, this.size);
                offset += this.size;
                break;
            }
        }

        return offset;
    }

    public boolean isWriteable()
    {
        return (this.flags & Flag_Write) != 0;
    }

    public boolean isReadable()
    {
        return (this.flags & Flag_Read) != 0;
    }

    public boolean isSubscribed()
    {
        return (this.flags & Flag_Subscribed) != 0;
    }

    public int size()
    {
        return this.size;
    }

    public int setValue(byte[] data, int dataOffset)
    {
        int offset = 0;

        switch (this.type)
        {
            case Integer: {
                this.value = ByteUtils.toInt32(data, dataOffset);
                offset += this.size();
                break;
            }

            case Float: {
                this.value = ByteUtils.toFloat(data, dataOffset);
                offset += this.size();
                break;
            }

            case Bool: {
                this.value = ByteUtils.toBool(data, dataOffset);
                offset += this.size();
                break;
            }

            case Data: {
                this.size = data[dataOffset];
                offset++;
                this.value = new byte[this.size];
                ArrayUtils.Copy(data, dataOffset +  1, this.value, 0, size);
                offset += this.size;
                break;
            }
        }

        return offset;
    }

    public int getValue(byte[] data, int offset)
    {
        switch (this.type)
        {
            case Integer: {
                return ByteUtils.fromInt32((int)this.value, data, offset);
            }

            case Float: {
                return ByteUtils.fromFloat((float)this.value, data, offset);
            }

            case Bool: {
                return ByteUtils.fromBool((boolean)this.value, data, offset);
            }

            case Data: {
                int size = this.size();
                data[offset] = (byte)size;
                ArrayUtils.Copy(data, offset, this.value, offset + 1, size);
                return 1 + size;
            }
        }

        return 0;
    }

    public int getInt()
    {
        return (int)this.value;
    }
    public float getFloat()
    {
        return (float)this.value;
    }
    public boolean getBool()
    {
        return (boolean)this.value;
    }

    public String getString()
    {
        return StringUtils.BytesToHex((byte[])this.value);
    }
}
